File: practice/python/trie/short_encoding_of_words.py
'''
- Problem: https://leetcode.com/problems/short-encoding-of-words/
- Solution:
  - Using trie data structure to manage next word is included by previous words or not.
  - Preprocessing: sort words descent size of each word.
  - Time and space complexity.
    - TC: O (nlog(n) + n*m) // m is size of word
    - SC: O(m)
'''

class Trie:
    def __init__(self):
        self.is_end = False
        self.next = [None]*27

def isInclude(t : Trie, word: str) -> bool:
    for i in reversed(range(len(word))):
        c = word[i]
        if t.next[ord(c) - ord('a')] == None:
            return False
        t = t.next[ord(c) - ord('a')]
    return True
    
def addWord(t: Trie, word: str):
    for i in reversed(range(len(word))):
        c = word[i]
        if t.next[ord(c) - ord('a')] == None:
            t.next[ord(c) - ord('a')] = Trie()
            
        t = t.next[ord(c) - ord('a')]
    t.is_true = True
    

def minimumLengthEncoding(words: list) -> int:
    words.sort(key = lambda x: -len(x))    
    res = 0
    t = Trie()
    
    for word in words:
        if not isInclude(t, word):
            addWord(t, word)
            res += len(word)+1    
    
    return res

if __name__ == "__main__":
    words = ["time", "me", "bell"]
    print(minimumLengthEncoding(words))

File: practice/cpp/trie/short_encoding_of_words.cpp
/*
- Problem: https://leetcode.com/problems/short-encoding-of-words/
- Solution:
  - Using trie data structure to manage next word is included by previous words or not.
  - Preprocessing: sort words descent size of each word.
*/

#include <iostream>
#include <string>
#include <vector>
#include <cstring>
#include <algorithm>
using namespace std;

struct Trie {
    bool is_end;
    Trie* next[27];
    
    Trie() {
        is_end = false;
        memset(next, 0x00,sizeof(Trie*)*27);
    }
};

void addWord(Trie* t, string word) {    
    for (int i = word.length()-1; i >= 0; i--) {
        char c = word[i];
        if (t->next[c - 'a'] == nullptr) {
            t->next[c- 'a'] = new Trie();
        }
        t = t->next[c - 'a'];
    }
    t->is_end = true;
}
    
bool isInclude(Trie* t, string word) {
    for (int i = word.length()-1; i >= 0; i--) {
        char c = word[i];
        if (t->next[c - 'a'] == nullptr) {
            return false;
        }
        t = t->next[c - 'a'];
    }
    return true;
        
}
    
int minimumLengthEncoding(vector<string>& words) {
    sort(words.begin(), words.end(), [](auto a, auto b) {
                                         return a.length() > b.length();
                                     });
        
    int res = 0;
        
    Trie* t = new Trie();
        
    for (auto word : words) {
        if (!isInclude(t, word)) {
            addWord(t, word);
            res += word.length() + 1;
        }
            
    }
    return res;
}

int main(void)
{
    vector<string> words = {"time", "me", "bell"};
    cout << minimumLengthEncoding(words) << endl;
}
